src/cargo/ops/cargo_package.rs

```diff
@@ -1076,3 +1239,101 @@ fn check_filename(file: &Path, shell: &mut Shell) -> CargoResult<()> {
     }
     Ok(())
 }
+
+/// Manages a temporary local registry that we use to overlay our new packages on the
+/// upstream registry. This way we can build lockfiles that depend on the new packages even
+/// before they're published.
+struct TmpRegistry<'a> {
+    gctx: &'a GlobalContext,
+    upstream: SourceId,
+    root: Filesystem,
+}
+
+impl<'a> TmpRegistry<'a> {
+    fn new(gctx: &'a GlobalContext, root: Filesystem, upstream: SourceId) -> CargoResult<Self> {
+        // If there's an old temporary registry, delete it. (How does this interact with file locking?
+        // locks are only on files, not on directory trees...)
+        let path = root.as_path_unlocked();
+        if path.exists() {
+            paths::remove_dir_all(path)?;
+        }
+        let slf = Self {
+            gctx,
+            root,
+            upstream,
+        };
+        Ok(slf)
+    }
+
+    fn index_path(&self) -> Filesystem {
+        self.root.join("index")
+    }
+
+    fn add_package(&mut self, package: &Package, tar: &FileLock) -> CargoResult<()> {
+        debug!(
+            "adding package {}@{} to local overlay at {}",
+            package.name(),
+            package.version(),
+            self.root.as_path_unlocked().display()
+        );
+        {
+            let mut tar_copy = self.root.open_rw_exclusive_create(
+                package.package_id().tarball_name(),
+                self.gctx,
+                "temporary package registry",
+            )?;
+            tar.file().seek(SeekFrom::Start(0))?;
+            std::io::copy(&mut tar.file(), &mut tar_copy)?;
+            tar_copy.flush()?;
+        }
+
+        let new_crate =
+            super::registry::prepare_transmit(self.gctx, package, self.upstream, false)?;
+
+        tar.file().seek(SeekFrom::Start(0))?;
+        let cksum = cargo_util::Sha256::new()
+            .update_file(tar.file())?
+            .finish_hex();
+
+        let deps: Vec<_> = new_crate
+            .deps
+            .into_iter()
+            .map(|dep| {
+                serde_json::json!({
+                    "name": dep.name,
+                    "req": dep.version_req,
+                    "features": dep.features,
+                    "default_features": dep.default_features,
+                    "target": dep.target,
+                    "optional": dep.optional,
+                    "kind": dep.kind,
+                    "registry": dep.registry,
+                    "artifact": dep.artifact,
+                    "bindep_target": dep.bindep_target,
+                    "lib": dep.lib,
+                })
+            })
+            .collect();
+
+        let index_line = serde_json::json!({
+            "name": new_crate.name,
+            "vers": new_crate.vers,
+            "deps": deps,
+            "cksum": cksum,
+            "features": new_crate.features,
+            "yanked": false,
+            "links": new_crate.links,
+            "v": 2,
+        })
+        .to_string();
```

Comment:
How about we use [IndexPackage](https://doc.rust-lang.org/nightly/nightly-rustc/cargo/sources/registry/index/struct.IndexPackage.html) for this?

Reply:
Thanks for the pointer! It's a little awkward because `NewCrate` and `IndexPackage` make different choices about `String` vs `InternedString`/`Cow`. But I guess the extra conversions are outweighed by the fact that it'll be easier to keep the index format in sync.